### PR TITLE
DD03 - Global configuration Monitor

### PR DESCRIPTION
**Pull request by yasmine9**
# DD03 - Global configuration Monitor

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [ ] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this agent please describe it here.

### CODE REVIEW

File: DYDX-Bots/Configuration-Monitor/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  - Type is always set to "Info".
+  - Metadata includes:
+    - `ConfigHash`: hash of the applied global configuration.
+
+- DYDX-3-3
+- Fired when `LogGlobalConfigurationRemoved` event is emitted on the perpetual contract.
+- Severity is always set to "Info".
+- Type is always set to "Info".
+- Metadata includes:
+  - `ConfigHash`: hash of the removed global configuration.
```
Comment (RCantu92) [inline DYDX-Bots/Configuration-Monitor/README.md:34]:
```suggestion
    - Fired when `LogGlobalConfigurationRemoved` event is emitted on the perpetual contract.
    - Severity is always set to "Info".
    - Type is always set to "Info".
    - Metadata includes:
      - `ConfigHash`: hash of the removed global configuration.
```
Missing indentation

File: DYDX-Bots/Configuration-Monitor/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+- 0x537eb31c5d265e25c993c75b9716042758a9fedfcd4dd47e8439936781f55fdc (`LogGlobalConfigurationApplied` event)
+
+### Ropsten Testnet
+
+> Transactions were generated through our PoC contract deployed on Ropsten testnet.
+> To test the transactions, `PERPETUAL_PROXY` should be changed to `TEST_PROXY` in `agent.ts` Line 51.
+
+- 0xced12eb03c0f5d3bc1f2c7fc341207b6b07ce7b80290bfe36118fb428d79de93 (`LogGlobalConfigurationRegistered` event)
+- 0x3e3edfdb159f1083e3f0a05e69ca53661ad8948a39eec62908c74fa3b4703e91 (`LogGlobalConfigurationApplied` event)
+- 0x0f1f9946ea28257c94bc312b71890759e553b6ff33786e60e2e07cb75c290635 (`LogGlobalConfigurationRemoved` event)
```
Comment (RCantu92) [inline DYDX-Bots/Configuration-Monitor/README.md:52]: Again, could get these transactions to generate a finding. Interestingly, the mainnet transactions did generate a finding.

**Review by RCantu92 — COMMENTED** [review]
In addition to the smal suggestions here, the same comments from the "Frozen state" bot's `README` apply here. (i.e. verifying the contract on etherscan and the testnet transactions not generating findings.)

Outside of that, great job 👍

**Review by RCantu92 — APPROVED** [review]
Went back and was able to generate the testnet findings.

Approved ✅

File: DYDX-Bots/Configuration-Monitor/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          : "LogGlobalConfigurationApplied"
+          ? "applied"
+          : "removed";
+
+      const alertId =
+        log.name === "LogGlobalConfigurationRegistered"
+          ? "DYDX-3-1"
+          : "LogGlobalConfigurationApplied"
+          ? "DYDX-3-2"
+          : "DYDX-3-3";
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply DYDX-Bots/Configuration-Monitor/src/agent.ts:29]: Yes, that was changed automatically by the formatter.

File: DYDX-Bots/Configuration-Monitor/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      configHash: configHash,
+    },
+  });
+};
+
+describe("Global configuration monitor tests suite", () => {
+  const perpetual = createAddress("0x1");
+  const PERPETUAL_IFACE = new Interface(MONITORED_EVENTS);
+
+  const mockProvider = {
```
Comment (RCantu92) [inline DYDX-Bots/Configuration-Monitor/src/agent.spec.ts:39]:
```suggestion
  const mockNetworkManager = {
```

File: DYDX-Bots/Configuration-Monitor/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        createAddress("0xa1"),
+        log3.data,
+        ...log3.topics
+      );
+
+    const findings: Finding[] = await handler(tx);
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should ignore other events on pereptual contract", async () => {
```
Comment (RCantu92) [inline DYDX-Bots/Configuration-Monitor/src/agent.spec.ts:87]:
```suggestion
  it("should ignore other events on perpetual contract", async () => {
```

**Review by RCantu92 — APPROVED** [review]
Just a couple of minor things. Overall, LGTM. Approved ✅

**Pull request merged**